Test failure with Python 3.9

Running the test suite og compyle 0.7 with Python 3.9 I get 6 errors.
Looking at the release notes of Python 3.9 I see that a new parser has been introduced but, as far as I can understand, the `ast` module should not be affected by the change.

**Update**: in [1] it is suggested to ``export PYTHONOLDPARSER=1`` to force the use of the old parser.
I can confirm that anso in this case I get the same results.

[1] https://docs.python.org/3/whatsnew/3.9.html#new-parser

```
============================= test session starts ==============================
platform linux -- Python 3.9.0, pytest-4.6.11, py-1.9.0, pluggy-0.13.0
rootdir: /build/compyle-0.7
collected 295 items / 4 deselected / 291 selected

compyle/tests/test_array.py .ss.ss.ss.ss.ss.ss.ss.ss.ss.ss.ss.ss.ss..ss. [ 15%]
ss.ss                                                                    [ 16%]
compyle/tests/test_ast_utils.py .......                                  [ 19%]
compyle/tests/test_capture_stream.py .....                               [ 20%]
compyle/tests/test_config.py ............                                [ 25%]
compyle/tests/test_cython_generator.py .................                 [ 30%]
compyle/tests/test_ext_module.py ......                                  [ 32%]
compyle/tests/test_gpu_struct.py s                                       [ 33%]
compyle/tests/test_jit.py .......................                        [ 41%]
compyle/tests/test_low_level.py ssss....                                 [ 43%]
compyle/tests/test_parallel.py s..ss.s.ss.s.ss.ss.ss.sss..ss..ss..ss..ss [ 58%]
..ss..ss.s.ss.s.ss.ss.ss.sss..ss..ss..ss..ss..s                          [ 74%]
compyle/tests/test_profile.py ..s                                        [ 75%]
compyle/tests/test_template.py .....                                     [ 76%]
compyle/tests/test_translator.py ...........F.....F...................F. [ 90%]
......FF.F....                                                           [ 95%]
compyle/tests/test_transpiler.py ...                                     [ 96%]
compyle/tests/test_types.py .......                                      [ 98%]
compyle/tests/test_utils.py ....                                         [100%]

=================================== FAILURES ===================================
________________________________ test_subscript ________________________________

    def test_subscript():
        # Given
        src = dedent('''
        x[1]
        ''')
    
        # When
>       code = py2c(src)

compyle/tests/test_translator.py:282: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compyle/translator.py:65: in py2c
    result = converter.convert(src)
compyle/translator.py:252: in convert
    result = self.visit(code)
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:652: in visit_Module
    return '\n'.join(
compyle/translator.py:653: in <genexpr>
    self.visit(item) for item in node.body
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:438: in visit_Expr
    return self.visit(node.value) + ';'
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:704: in visit_Subscript
    self.visit(node.value), self.visit(node.slice.value)
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
/usr/lib/python3.9/ast.py:411: in generic_visit
    for field, value in iter_fields(node):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

node = 1

    def iter_fields(node):
        """
        Yield a tuple of ``(fieldname, value)`` for each field in ``node._fields``
        that is present on *node*.
        """
>       for field in node._fields:
E       AttributeError: 'int' object has no attribute '_fields'

/usr/lib/python3.9/ast.py:249: AttributeError
___________________________ test_annotated_function ____________________________

    def test_annotated_function():
        # Given/When
        t = CConverter()
>       code = t.parse_function(annotated_f)

compyle/tests/test_translator.py:411: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compyle/translator.py:307: in parse_function
    code = self.convert(src)
compyle/translator.py:252: in convert
    result = self.visit(code)
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:652: in visit_Module
    return '\n'.join(
compyle/translator.py:653: in <genexpr>
    self.visit(item) for item in node.body
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:584: in visit_FunctionDef
    body = '\n'.join(self._indent_block(self.visit(item))
compyle/translator.py:584: in <genexpr>
    body = '\n'.join(self._indent_block(self.visit(item))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:692: in visit_Return
    return 'return %s;' % (self.visit(node.value))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <compyle.translator.CConverter object at 0x7fece8274520>
node = <ast.Subscript object at 0x7fece8274610>

    def visit_Subscript(self, node):
        return '%s[%s]' % (
>           self.visit(node.value), self.visit(node.slice.value)
        )
E       AttributeError: 'Name' object has no attribute 'value'

compyle/translator.py:704: AttributeError
_____________________________ test_declare_matrix ______________________________

    def test_declare_matrix():
        # Given
        src = dedent('''
        x = declare('matrix((3,))')
        do(x[0])
        ''')
    
        # When
>       code = py2c(src)

compyle/tests/test_translator.py:950: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compyle/translator.py:65: in py2c
    result = converter.convert(src)
compyle/translator.py:252: in convert
    result = self.visit(code)
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:652: in visit_Module
    return '\n'.join(
compyle/translator.py:653: in <genexpr>
    self.visit(item) for item in node.body
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:438: in visit_Expr
    return self.visit(node.value) + ';'
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:392: in visit_Call
    args=', '.join(self.visit(x) for x in node.args)
compyle/translator.py:392: in <genexpr>
    args=', '.join(self.visit(x) for x in node.args)
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:704: in visit_Subscript
    self.visit(node.value), self.visit(node.slice.value)
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
/usr/lib/python3.9/ast.py:411: in generic_visit
    for field, value in iter_fields(node):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

node = 0

    def iter_fields(node):
        """
        Yield a tuple of ``(fieldname, value)`` for each field in ``node._fields``
        that is present on *node*.
        """
>       for field in node._fields:
E       AttributeError: 'int' object has no attribute '_fields'

/usr/lib/python3.9/ast.py:249: AttributeError
_____________________________ test_cuda_conversion _____________________________

    def test_cuda_conversion():
>       check_opencl_cuda_conversion(CUDAConverter)

compyle/tests/test_translator.py:1218: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compyle/tests/test_translator.py:1204: in check_opencl_cuda_conversion
    code = converter.convert(src)
compyle/translator.py:252: in convert
    result = self.visit(code)
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:652: in visit_Module
    return '\n'.join(
compyle/translator.py:653: in <genexpr>
    self.visit(item) for item in node.body
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:584: in visit_FunctionDef
    body = '\n'.join(self._indent_block(self.visit(item))
compyle/translator.py:584: in <genexpr>
    body = '\n'.join(self._indent_block(self.visit(item))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:355: in visit_Assign
    return '%s = %s;' % (self.visit(left), self.visit(right))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <compyle.translator.CUDAConverter object at 0x7fece768c850>
node = <ast.Subscript object at 0x7fece7827d90>

    def visit_Subscript(self, node):
        return '%s[%s]' % (
>           self.visit(node.value), self.visit(node.slice.value)
        )
E       AttributeError: 'Name' object has no attribute 'value'

compyle/translator.py:704: AttributeError
____________________________ test_opencl_conversion ____________________________

    def test_opencl_conversion():
>       check_opencl_cuda_conversion(OpenCLConverter)

compyle/tests/test_translator.py:1222: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compyle/tests/test_translator.py:1204: in check_opencl_cuda_conversion
    code = converter.convert(src)
compyle/translator.py:252: in convert
    result = self.visit(code)
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:652: in visit_Module
    return '\n'.join(
compyle/translator.py:653: in <genexpr>
    self.visit(item) for item in node.body
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:584: in visit_FunctionDef
    body = '\n'.join(self._indent_block(self.visit(item))
compyle/translator.py:584: in <genexpr>
    body = '\n'.join(self._indent_block(self.visit(item))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:355: in visit_Assign
    return '%s = %s;' % (self.visit(left), self.visit(right))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <compyle.translator.OpenCLConverter object at 0x7fece1908310>
node = <ast.Subscript object at 0x7fece19082b0>

    def visit_Subscript(self, node):
        return '%s[%s]' % (
>           self.visit(node.value), self.visit(node.slice.value)
        )
E       AttributeError: 'Name' object has no attribute 'value'

compyle/translator.py:704: AttributeError
__________________________ test_cuda_local_conversion __________________________

    def test_cuda_local_conversion():
        @annotate(xc='ldoublep', yc='lintp')
        def knl(xc, yc):
            xc[LID_0] = 1
            yc[LID_0] = 1
    
        # When
        converter = CUDAConverter()
>       code = converter.parse(knl)

compyle/tests/test_translator.py:1254: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compyle/translator.py:281: in parse
    code = self.parse_function(obj, declarations=declarations)
compyle/translator.py:307: in parse_function
    code = self.convert(src)
compyle/translator.py:252: in convert
    result = self.visit(code)
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:652: in visit_Module
    return '\n'.join(
compyle/translator.py:653: in <genexpr>
    self.visit(item) for item in node.body
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:584: in visit_FunctionDef
    body = '\n'.join(self._indent_block(self.visit(item))
compyle/translator.py:584: in <genexpr>
    body = '\n'.join(self._indent_block(self.visit(item))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:355: in visit_Assign
    return '%s = %s;' % (self.visit(left), self.visit(right))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <compyle.translator.CUDAConverter object at 0x7fece7b86640>
node = <ast.Subscript object at 0x7fece80d05b0>

    def visit_Subscript(self, node):
        return '%s[%s]' % (
>           self.visit(node.value), self.visit(node.slice.value)
        )
E       AttributeError: 'Name' object has no attribute 'value'

compyle/translator.py:704: AttributeError
=============================== warnings summary ===============================
compyle/parallel.py:928
  /build/compyle-0.7/.pybuild/cpython3_3.9_compyle/build/compyle/parallel.py:928: SyntaxWarning: "is" with a literal. Did you mean "=="?
    if func_type is 'input':

compyle/parallel.py:1035
  /build/compyle-0.7/.pybuild/cpython3_3.9_compyle/build/compyle/parallel.py:1035: SyntaxWarning: "is" with a literal. Did you mean "=="?
    if func_type is 'input':

compyle/parallel.py:1150
  /build/compyle-0.7/.pybuild/cpython3_3.9_compyle/build/compyle/parallel.py:1150: SyntaxWarning: "is" with a literal. Did you mean "=="?
    return_type = 'void' if func_type is 'output' else self.type

.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_array.py::test_radix_sort_by_keys
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_array.py::test_radix_sort_by_keys
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_array.py::test_radix_sort_by_keys
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_array.py::test_radix_sort_by_keys
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_jit.py::TestAnnotationHelper::test_const_as_call_arg
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_jit.py::TestAnnotationHelper::test_const_as_call_arg
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_jit.py::TestAnnotationHelper::test_const_as_call_arg
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_jit.py::TestAnnotationHelper::test_const_in_return
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_jit.py::TestAnnotationHelper::test_const_in_return
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_jit.py::TestAnnotationHelper::test_const_in_return
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_jit.py::TestAnnotationHelper::test_undeclared_variable_declaration_in_for
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_parallel.py::TestParallelUtilsJIT::test_reduction_works_with_external_func_cython
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_parallel.py::TestParallelUtilsJIT::test_reduction_works_with_external_func_cython
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_parallel.py::TestParallelUtilsJIT::test_reduction_works_with_map_cython
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_parallel.py::TestParallelUtilsJIT::test_reduction_works_with_map_cython
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_parallel.py::TestParallelUtilsJIT::test_reduction_works_with_map_cython
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_parallel.py::TestParallelUtilsJIT::test_unique_scan_cython
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_parallel.py::TestParallelUtilsJIT::test_unique_scan_cython
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_parallel.py::TestParallelUtilsJIT::test_unique_scan_cython_parallel
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_parallel.py::TestParallelUtilsJIT::test_unique_scan_cython_parallel
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_simple_assignment_expression
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_simple_assignment_expression
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_simple_assignment_expression
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_simple_assignment_expression
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_multiple_assignment_expressions
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_multiple_assignment_expressions
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_multiple_assignment_expressions
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_multiple_assignment_expressions
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_multiple_assignment_expressions
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_if_block
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_if_block
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_if_block
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_if_block
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_if_block
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_conditionals
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_conditionals
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_conditionals
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_conditionals
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_conditionals
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_conditionals
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_conditionals
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_conditionals
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_conditionals
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_conditionals
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_conditionals
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_conditionals
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_ternary_operator
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_ternary_operator
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_ternary_operator
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_ternary_operator
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_multiple_boolops
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_multiple_boolops
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_multiple_boolops
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_multiple_boolops
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_multiple_boolops
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_multiple_bitwise_ops
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_multiple_bitwise_ops
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_multiple_bitwise_ops
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_power
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_power
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_calling_function
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_calling_function
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_calling_printf_with_string
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_calling_printf_with_string
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_simple_function_with_return
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_py3_annotations
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_calling_method_of_known_type
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_calling_method_of_known_type
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_calling_method_of_known_type_in_method
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_calling_method_of_known_type_in_method
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_while
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for_with_decreasing_range
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for_with_decreasing_range
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for_with_decreasing_range
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for_with_declare
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_two_fors
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_two_fors
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for_with_symbols
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for_with_symbols
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for_with_symbols
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_nested_for_with_symbols
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_nested_for_with_symbols
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_nested_for_with_symbols
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_nested_for_with_symbols
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_nested_for_with_symbols
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_nested_for_with_symbols
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for_with_break_continue
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for_with_break_continue
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for_with_break_continue
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for_with_break_continue
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_for_with_break_continue
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_attribute_access
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_declare_call_declares_variable
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_class
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_class
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_handles_parsing_functions
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_handles_parsing_functions
  /usr/lib/python3.9/ast.py:407: DeprecationWarning: visit_Num is deprecated; add visit_Constant
    return visitor(node)

.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_jit.py::TestAnnotationHelper::test_non_jit_call_as_call_arg
  /build/compyle-0.7/.pybuild/cpython3_3.9_compyle/build/compyle/jit.py:150: UserWarning: 
  In code in line 5:
  
      return g(sin(a))
               ^
  
  
  Function called is not marked by the annotate decorator. Argument
  type defaulting to 'double'. If the type is not 'double', store
  the value in a variable of appropriate type and use the variable
  
    warnings.warn(msg)

.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_calling_printf_with_string
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_calling_printf_with_string
  /usr/lib/python3.9/ast.py:407: DeprecationWarning: visit_Str is deprecated; add visit_Constant
    return visitor(node)

.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_bool_true_false_and_none
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_bool_true_false_and_none
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_bool_true_false_and_none
.pybuild/cpython3_3.9_compyle/build/compyle/tests/test_translator.py::test_bool_true_false_and_none
  /usr/lib/python3.9/ast.py:407: DeprecationWarning: visit_NameConstant is deprecated; add visit_Constant
    return visitor(node)

-- Docs: https://docs.pytest.org/en/latest/warnings.html
= 6 failed, 197 passed, 88 skipped, 4 deselected, 113 warnings in 147.22 seconds =
```

Tested on GNU/Linux, Debian (sid) x86 64bit.
numpy: 1.19.4
pyopencl:  2019.1.1-1

Please find below the test log (please note that the same code works without errors woth Python 3.8.

```
============================= test session starts ==============================
platform linux -- Python 3.9.0+, pytest-4.6.11, py-1.9.0, pluggy-0.13.0
rootdir: /build/compyle-0.7
collected 295 items / 4 deselected / 291 selected

compyle/tests/test_array.py ..s..s..s..s..s..s..s.Fs.Fs..s.Fs..s..s...s. [ 15%]
.s..s                                                                    [ 16%]
compyle/tests/test_ast_utils.py .......                                  [ 19%]
compyle/tests/test_capture_stream.py .....                               [ 20%]
compyle/tests/test_config.py ............                                [ 25%]
compyle/tests/test_cython_generator.py .................                 [ 30%]
compyle/tests/test_ext_module.py ......                                  [ 32%]
compyle/tests/test_gpu_struct.py s                                       [ 33%]
compyle/tests/test_jit.py .......................                        [ 41%]
compyle/tests/test_low_level.py s.s.....                                 [ 43%]
compyle/tests/test_parallel.py s...s.s....s..s..s..s..Fs...s...s...s...s [ 58%]
..Fs...s.s....s..s..s..s..Fs...s...s...s...s..F                          [ 74%]
compyle/tests/test_profile.py ...                                        [ 75%]
compyle/tests/test_template.py .....                                     [ 76%]
compyle/tests/test_translator.py ....................................... [ 90%]
..............                                                           [ 95%]
compyle/tests/test_transpiler.py ...                                     [ 96%]
compyle/tests/test_types.py .......                                      [ 98%]
compyle/tests/test_utils.py ....                                         [100%]

=================================== FAILURES ===================================
_____________________________ test_remove[opencl] ______________________________

obj = <compyle.array.Array object at 0x7fd68009f2e0>, args = (), kwargs = {}
key = ()

    def wrapper(obj, *args, **kwargs):
        if kwargs:
            key = (_HasKwargs, frozenset(six.iteritems(kwargs))) + args
        else:
            key = args
    
        try:
>           return getattr(obj, cache_dict_name)[key]
E           AttributeError: 'Array' object has no attribute '_memoize_dic__get_remove_kernels'

/usr/lib/python3/dist-packages/pytools/__init__.py:581: AttributeError

During handling of the above exception, another exception occurred:

obj = <pyopencl.elementwise.ElementwiseKernel object at 0x7fd66b8450d0>
args = (False,), kwargs = {}, key = (False,)

    def wrapper(obj, *args, **kwargs):
        if kwargs:
            key = (_HasKwargs, frozenset(six.iteritems(kwargs))) + args
        else:
            key = args
    
        try:
>           return getattr(obj, cache_dict_name)[key]
E           AttributeError: 'ElementwiseKernel' object has no attribute '_memoize_dic_get_kernel'

/usr/lib/python3/dist-packages/pytools/__init__.py:581: AttributeError

During handling of the above exception, another exception occurred:

backend = 'opencl'

    @check_all_backends
    def test_remove(backend):
        check_import(backend)
    
        # Given
        dev_array = Array(np.int32, backend=backend)
        orig_array = array.arange(0, 16, 1, dtype=np.int32,
                                  backend=backend)
        dev_array.set_data(orig_array)
        indices = array.arange(0, 8, 1, dtype=np.int32, backend=backend)
    
        # When
>       dev_array.remove(indices)

compyle/tests/test_array.py:153: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compyle/profile.py:48: in wrapper
    return method(*args, **kwargs)
compyle/array.py:913: in remove
    fill_if_remove_knl, remove_knl = self._get_remove_kernels()
/usr/lib/python3/dist-packages/pytools/__init__.py:583: in wrapper
    result = function(obj, *args, **kwargs)
compyle/array.py:879: in _get_remove_kernels
    fill_if_remove_knl = parallel.Elementwise(
compyle/parallel.py:561: in __init__
    self.elementwise = ElementwiseBase(func, backend=backend)
compyle/parallel.py:418: in __init__
    self.c_func = self._generate()
compyle/parallel.py:474: in _generate
    all_source = knl.get_kernel(False)[0].program.source
/usr/lib/python3/dist-packages/pytools/__init__.py:583: in wrapper
    result = function(obj, *args, **kwargs)
/usr/lib/python3/dist-packages/pyopencl/elementwise.py:221: in get_kernel
    knl, arg_descrs = get_elwise_kernel_and_types(
/usr/lib/python3/dist-packages/pyopencl/elementwise.py:158: in get_elwise_kernel_and_types
    prg = get_elwise_program(
/usr/lib/python3/dist-packages/pyopencl/elementwise.py:111: in get_elwise_program
    return Program(context, source).build(options)
/usr/lib/python3/dist-packages/pyopencl/__init__.py:506: in build
    self._prg, was_cached = self._build_and_catch_errors(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pyopencl.Program object at 0x7fd66b845640>
build_func = <function Program.build.<locals>.<lambda> at 0x7fd66ae6e550>
options_bytes = b'-I /usr/lib/python3/dist-packages/pyopencl/cl'
source = '//CL//\n        \n#define local_barrier() barrier(CLK_LOCAL_MEM_FENCE);\n\n#define WITHIN_KERNEL /* empty */\n#define...      {\n            fill_if_remove(i, indices, if_remove);\n          }\n          \n          ;\n        }\n        '

    def _build_and_catch_errors(self, build_func, options_bytes, source=None):
        try:
            return build_func()
        except _cl.RuntimeError as e:
            msg = str(e)
            if options_bytes:
                msg = msg + "\n(options: %s)" % options_bytes.decode("utf-8")
    
            if source is not None:
                from tempfile import NamedTemporaryFile
                srcfile = NamedTemporaryFile(mode="wt", delete=False, suffix=".cl")
                try:
                    srcfile.write(source)
                finally:
                    srcfile.close()
    
                msg = msg + "\n(source saved as %s)" % srcfile.name
    
            code = e.code
            routine = e.routine
    
            err = _cl.RuntimeError(
                    _cl._ErrorRecord(
                        msg=msg,
                        code=code,
                        routine=routine))
    
        # Python 3.2 outputs the whole list of currently active exceptions
        # This serves to remove one (redundant) level from that nesting.
>       raise err
E       pyopencl._cl.RuntimeError: clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE
E       
E       Build on <pyopencl.Device 'pthread-Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz' on 'Portable Computing Language' at 0x2b76ad0>:
E       
E       error: /build/compyle-0.7/.pybuild/cpython3_3.9_compyle/.cache/pocl/kcache/tempfile-4b-23-be-9e-4f.cl:74:14: array subscript is not an integer
E       warning: /build/compyle-0.7/.pybuild/cpython3_3.9_compyle/.cache/pocl/kcache/tempfile-4b-23-be-9e-4f.cl:62:9: 'max' macro redefined
E       warning: /build/compyle-0.7/.pybuild/cpython3_3.9_compyle/.cache/pocl/kcache/tempfile-4b-23-be-9e-4f.cl:62:9: 'max' macro redefined
E       
E       (options: -I /usr/lib/python3/dist-packages/pyopencl/cl)
E       (source saved as /tmp/tmp6ewsokv5.cl)

/usr/lib/python3/dist-packages/pyopencl/__init__.py:554: RuntimeError
----------------------------- Captured stderr call -----------------------------
1 warning generated.
2 warnings and 1 error generated.
______________________________ test_align[opencl] ______________________________

obj = <pyopencl.elementwise.ElementwiseKernel object at 0x7fd67819cf40>
args = (False,), kwargs = {}, key = (False,)

    def wrapper(obj, *args, **kwargs):
        if kwargs:
            key = (_HasKwargs, frozenset(six.iteritems(kwargs))) + args
        else:
            key = args
    
        try:
>           return getattr(obj, cache_dict_name)[key]
E           AttributeError: 'ElementwiseKernel' object has no attribute '_memoize_dic_get_kernel'

/usr/lib/python3/dist-packages/pytools/__init__.py:581: AttributeError

During handling of the above exception, another exception occurred:

backend = 'opencl'

    @check_all_backends
    def test_align(backend):
        check_import(backend)
    
        # Given
        dev_array = Array(np.int32, backend=backend)
        orig_array = array.arange(0, 16, 1, dtype=np.int32, backend=backend)
        dev_array.set_data(orig_array)
        indices = array.arange(15, -1, -1, dtype=np.int32, backend=backend)
    
        # When
>       dev_array = dev_array.align(indices)

compyle/tests/test_array.py:170: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compyle/array.py:922: in align
    return take(self, indices, backend=self.backend, out=out)
compyle/array.py:591: in take
    take_knl(indices, ary, out)
compyle/parallel.py:573: in __call__
    self.elementwise(*args, **kwargs)
compyle/profile.py:48: in wrapper
    return method(*args, **kwargs)
compyle/jit.py:334: in __call__
    c_func = self._generate_kernel(*args)
compyle/jit.py:29: in wrapper
    setattr(f, 'cached_kernel', {key_val: method(*args)})
compyle/jit.py:322: in _generate_kernel
    return self._generate(declarations=declarations)
compyle/parallel.py:474: in _generate
    all_source = knl.get_kernel(False)[0].program.source
/usr/lib/python3/dist-packages/pytools/__init__.py:583: in wrapper
    result = function(obj, *args, **kwargs)
/usr/lib/python3/dist-packages/pyopencl/elementwise.py:221: in get_kernel
    knl, arg_descrs = get_elwise_kernel_and_types(
/usr/lib/python3/dist-packages/pyopencl/elementwise.py:158: in get_elwise_kernel_and_types
    prg = get_elwise_program(
/usr/lib/python3/dist-packages/pyopencl/elementwise.py:111: in get_elwise_program
    return Program(context, source).build(options)
/usr/lib/python3/dist-packages/pyopencl/__init__.py:506: in build
    self._prg, was_cached = self._build_and_catch_errors(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pyopencl.Program object at 0x7fd67819c700>
build_func = <function Program.build.<locals>.<lambda> at 0x7fd66aee78b0>
options_bytes = b'-I /usr/lib/python3/dist-packages/pyopencl/cl'
source = '//CL//\n        \n#define local_barrier() barrier(CLK_LOCAL_MEM_FENCE);\n\n#define WITHIN_KERNEL /* empty */\n#define...      {\n            take_elwise(i, indices, ary, out_ary);\n          }\n          \n          ;\n        }\n        '

    def _build_and_catch_errors(self, build_func, options_bytes, source=None):
        try:
            return build_func()
        except _cl.RuntimeError as e:
            msg = str(e)
            if options_bytes:
                msg = msg + "\n(options: %s)" % options_bytes.decode("utf-8")
    
            if source is not None:
                from tempfile import NamedTemporaryFile
                srcfile = NamedTemporaryFile(mode="wt", delete=False, suffix=".cl")
                try:
                    srcfile.write(source)
                finally:
                    srcfile.close()
    
                msg = msg + "\n(source saved as %s)" % srcfile.name
    
            code = e.code
            routine = e.routine
    
            err = _cl.RuntimeError(
                    _cl._ErrorRecord(
                        msg=msg,
                        code=code,
                        routine=routine))
    
        # Python 3.2 outputs the whole list of currently active exceptions
        # This serves to remove one (redundant) level from that nesting.
>       raise err
E       pyopencl._cl.RuntimeError: clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE
E       
E       Build on <pyopencl.Device 'pthread-Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz' on 'Portable Computing Language' at 0x2b76ad0>:
E       
E       error: /build/compyle-0.7/.pybuild/cpython3_3.9_compyle/.cache/pocl/kcache/tempfile-99-01-74-f4-2f.cl:74:21: array subscript is not an integer
E       warning: /build/compyle-0.7/.pybuild/cpython3_3.9_compyle/.cache/pocl/kcache/tempfile-99-01-74-f4-2f.cl:62:9: 'max' macro redefined
E       warning: /build/compyle-0.7/.pybuild/cpython3_3.9_compyle/.cache/pocl/kcache/tempfile-99-01-74-f4-2f.cl:62:9: 'max' macro redefined
E       
E       (options: -I /usr/lib/python3/dist-packages/pyopencl/cl)
E       (source saved as /tmp/tmpnzzudb55.cl)

/usr/lib/python3/dist-packages/pyopencl/__init__.py:554: RuntimeError
----------------------------- Captured stderr call -----------------------------
1 warning generated.
2 warnings and 1 error generated.
___________________________ test_copy_values[opencl] ___________________________

obj = <pyopencl.elementwise.ElementwiseKernel object at 0x7fd66be8e8e0>
args = (False,), kwargs = {}, key = (False,)

    def wrapper(obj, *args, **kwargs):
        if kwargs:
            key = (_HasKwargs, frozenset(six.iteritems(kwargs))) + args
        else:
            key = args
    
        try:
>           return getattr(obj, cache_dict_name)[key]
E           AttributeError: 'ElementwiseKernel' object has no attribute '_memoize_dic_get_kernel'

/usr/lib/python3/dist-packages/pytools/__init__.py:581: AttributeError

During handling of the above exception, another exception occurred:

backend = 'opencl'

    @check_all_backends
    def test_copy_values(backend):
        check_import(backend)
    
        # Given
        dev_array = make_dev_array(backend)
        dev_array.fill(2)
    
        dest = array.empty(8, dtype=np.int32, backend=backend)
        indices = array.arange(0, 8, 1, dtype=np.int32, backend=backend)
    
        # When
>       dev_array.copy_values(indices, dest)

compyle/tests/test_array.py:229: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compyle/array.py:933: in copy_values
    dest.dev[:len(indices.dev)] = take(
compyle/array.py:591: in take
    take_knl(indices, ary, out)
compyle/parallel.py:573: in __call__
    self.elementwise(*args, **kwargs)
compyle/profile.py:48: in wrapper
    return method(*args, **kwargs)
compyle/jit.py:334: in __call__
    c_func = self._generate_kernel(*args)
compyle/jit.py:29: in wrapper
    setattr(f, 'cached_kernel', {key_val: method(*args)})
compyle/jit.py:322: in _generate_kernel
    return self._generate(declarations=declarations)
compyle/parallel.py:474: in _generate
    all_source = knl.get_kernel(False)[0].program.source
/usr/lib/python3/dist-packages/pytools/__init__.py:583: in wrapper
    result = function(obj, *args, **kwargs)
/usr/lib/python3/dist-packages/pyopencl/elementwise.py:221: in get_kernel
    knl, arg_descrs = get_elwise_kernel_and_types(
/usr/lib/python3/dist-packages/pyopencl/elementwise.py:158: in get_elwise_kernel_and_types
    prg = get_elwise_program(
/usr/lib/python3/dist-packages/pyopencl/elementwise.py:111: in get_elwise_program
    return Program(context, source).build(options)
/usr/lib/python3/dist-packages/pyopencl/__init__.py:506: in build
    self._prg, was_cached = self._build_and_catch_errors(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pyopencl.Program object at 0x7fd678091f40>
build_func = <function Program.build.<locals>.<lambda> at 0x7fd66b85d040>
options_bytes = b'-I /usr/lib/python3/dist-packages/pyopencl/cl'
source = '//CL//\n        \n#define local_barrier() barrier(CLK_LOCAL_MEM_FENCE);\n\n#define WITHIN_KERNEL /* empty */\n#define...      {\n            take_elwise(i, indices, ary, out_ary);\n          }\n          \n          ;\n        }\n        '

    def _build_and_catch_errors(self, build_func, options_bytes, source=None):
        try:
            return build_func()
        except _cl.RuntimeError as e:
            msg = str(e)
            if options_bytes:
                msg = msg + "\n(options: %s)" % options_bytes.decode("utf-8")
    
            if source is not None:
                from tempfile import NamedTemporaryFile
                srcfile = NamedTemporaryFile(mode="wt", delete=False, suffix=".cl")
                try:
                    srcfile.write(source)
                finally:
                    srcfile.close()
    
                msg = msg + "\n(source saved as %s)" % srcfile.name
    
            code = e.code
            routine = e.routine
    
            err = _cl.RuntimeError(
                    _cl._ErrorRecord(
                        msg=msg,
                        code=code,
                        routine=routine))
    
        # Python 3.2 outputs the whole list of currently active exceptions
        # This serves to remove one (redundant) level from that nesting.
>       raise err
E       pyopencl._cl.RuntimeError: clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE
E       
E       Build on <pyopencl.Device 'pthread-Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz' on 'Portable Computing Language' at 0x2b76ad0>:
E       
E       error: /build/compyle-0.7/.pybuild/cpython3_3.9_compyle/.cache/pocl/kcache/tempfile-a1-64-a7-51-b4.cl:74:21: array subscript is not an integer
E       warning: /build/compyle-0.7/.pybuild/cpython3_3.9_compyle/.cache/pocl/kcache/tempfile-a1-64-a7-51-b4.cl:62:9: 'max' macro redefined
E       warning: /build/compyle-0.7/.pybuild/cpython3_3.9_compyle/.cache/pocl/kcache/tempfile-a1-64-a7-51-b4.cl:62:9: 'max' macro redefined
E       
E       (options: -I /usr/lib/python3/dist-packages/pyopencl/cl)
E       (source saved as /tmp/tmppoyaltma.cl)

/usr/lib/python3/dist-packages/pyopencl/__init__.py:554: RuntimeError
----------------------------- Captured stderr call -----------------------------
1 warning generated.
2 warnings and 1 error generated.
________ TestParallelUtils.test_repeated_scans_with_different_settings _________

self = <compyle.tests.test_parallel.TestParallelUtils testMethod=test_repeated_scans_with_different_settings>

    def test_repeated_scans_with_different_settings(self):
        importorskip('pyopencl')
        with use_config(use_double=False):
>           self._test_unique_scan(backend='opencl')

compyle/tests/test_parallel.py:197: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compyle/tests/test_parallel.py:436: in _test_unique_scan
    scan = Scan(input_f, output_f, 'a+b', dtype=np.int32, backend=backend)
compyle/parallel.py:1220: in __init__
    self.scan = ScanBase(input=input, output=output,
compyle/parallel.py:864: in __init__
    self.c_func = self._generate()
compyle/parallel.py:902: in _generate
    return self._generate_opencl_kernel(declarations=declarations)
compyle/parallel.py:1087: in _generate_opencl_kernel
    segment_expr, preamble = self._get_opencl_cuda_code(
compyle/parallel.py:1057: in _get_opencl_cuda_code
    self._wrap_ocl_function(self.input_func, func_type='input',
compyle/parallel.py:1022: in _wrap_ocl_function
    self.tp.add(func, declarations=declarations)
compyle/transpiler.py:278: in add
    code = self._cgen.parse(
compyle/translator.py:281: in parse
    code = self.parse_function(obj, declarations=declarations)
compyle/translator.py:307: in parse_function
    code = self.convert(src)
compyle/translator.py:252: in convert
    result = self.visit(code)
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:652: in visit_Module
    return '\n'.join(
compyle/translator.py:653: in <genexpr>
    self.visit(item) for item in node.body
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:584: in visit_FunctionDef
    body = '\n'.join(self._indent_block(self.visit(item))
compyle/translator.py:584: in <genexpr>
    body = '\n'.join(self._indent_block(self.visit(item))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:615: in visit_If
    cond=self.visit(node.test),
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:376: in visit_BoolOp
    return '(%s)' % (op.join(self.visit(x) for x in node.values))
compyle/translator.py:376: in <genexpr>
    return '(%s)' % (op.join(self.visit(x) for x in node.values))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:426: in visit_Compare
    self.visit(node.comparators[0]))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <compyle.translator.OpenCLConverter object at 0x7fd66bcbaa30>
node = <ast.Subscript object at 0x7fd66a950100>

    def visit_Subscript(self, node):
        if isinstance(node.slice, ast.Constant) or isinstance(node.slice, ast.Name):
            return '%s[%s]' % (
                self.visit(node.value), self.visit(node.slice)
            )
        else:
            return '%s[%s]' % (
>               self.visit(node.value), self.visit(node.slice.value)
            )
E           AttributeError: 'BinOp' object has no attribute 'value'

compyle/translator.py:709: AttributeError
__________________ TestParallelUtils.test_unique_scan_opencl ___________________

self = <compyle.tests.test_parallel.TestParallelUtils testMethod=test_unique_scan_opencl>

    def test_unique_scan_opencl(self):
        importorskip('pyopencl')
>       self._test_unique_scan(backend='opencl')

compyle/tests/test_parallel.py:134: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compyle/tests/test_parallel.py:436: in _test_unique_scan
    scan = Scan(input_f, output_f, 'a+b', dtype=np.int32, backend=backend)
compyle/parallel.py:1220: in __init__
    self.scan = ScanBase(input=input, output=output,
compyle/parallel.py:864: in __init__
    self.c_func = self._generate()
compyle/parallel.py:902: in _generate
    return self._generate_opencl_kernel(declarations=declarations)
compyle/parallel.py:1087: in _generate_opencl_kernel
    segment_expr, preamble = self._get_opencl_cuda_code(
compyle/parallel.py:1057: in _get_opencl_cuda_code
    self._wrap_ocl_function(self.input_func, func_type='input',
compyle/parallel.py:1022: in _wrap_ocl_function
    self.tp.add(func, declarations=declarations)
compyle/transpiler.py:278: in add
    code = self._cgen.parse(
compyle/translator.py:281: in parse
    code = self.parse_function(obj, declarations=declarations)
compyle/translator.py:307: in parse_function
    code = self.convert(src)
compyle/translator.py:252: in convert
    result = self.visit(code)
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:652: in visit_Module
    return '\n'.join(
compyle/translator.py:653: in <genexpr>
    self.visit(item) for item in node.body
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:584: in visit_FunctionDef
    body = '\n'.join(self._indent_block(self.visit(item))
compyle/translator.py:584: in <genexpr>
    body = '\n'.join(self._indent_block(self.visit(item))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:615: in visit_If
    cond=self.visit(node.test),
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:376: in visit_BoolOp
    return '(%s)' % (op.join(self.visit(x) for x in node.values))
compyle/translator.py:376: in <genexpr>
    return '(%s)' % (op.join(self.visit(x) for x in node.values))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:426: in visit_Compare
    self.visit(node.comparators[0]))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <compyle.translator.OpenCLConverter object at 0x7fd66aff5760>
node = <ast.Subscript object at 0x7fd678071370>

    def visit_Subscript(self, node):
        if isinstance(node.slice, ast.Constant) or isinstance(node.slice, ast.Name):
            return '%s[%s]' % (
                self.visit(node.value), self.visit(node.slice)
            )
        else:
            return '%s[%s]' % (
>               self.visit(node.value), self.visit(node.slice.value)
            )
E           AttributeError: 'BinOp' object has no attribute 'value'

compyle/translator.py:709: AttributeError
_______ TestParallelUtilsJIT.test_repeated_scans_with_different_settings _______

func = <function ScanJIT._generate_kernel at 0x7fd68bf06430>
args = (<compyle.jit.ScanJIT object at 0x7fd66bcf22e0>,)
kwargs = {'ary': <compyle.array.Array object at 0x7fd66bcf2550>, 'unique': <compyle.array.Array object at 0x7fd66bcf28b0>, 'unique_count': <compyle.array.Array object at 0x7fd66bcf2820>}
key = ('gintp', 'gintp', 'gintp', <function TestParallelUtilsJIT._test_unique_scan.<locals>.input_f at 0x7fd66b3aa790>, <function TestParallelUtilsJIT._test_unique_scan.<locals>.output_f at 0x7fd66b384550>, 'a+b', ...)

    @my_decorator
    def _deco(func, *args, **kwargs):
        # by Michele Simionato
        # http://www.phyast.pitt.edu/~micheles/python/
        key = key_func(*args, **kwargs)
        try:
>           return func._memoize_dic[key]  # pylint: disable=protected-access
E           KeyError: ('gintp', 'gintp', 'gintp', <function TestParallelUtilsJIT._test_unique_scan.<locals>.input_f at 0x7fd66b3aa790>, <function TestParallelUtilsJIT._test_unique_scan.<locals>.output_f at 0x7fd66b384550>, 'a+b', 'opencl', False, False)

/usr/lib/python3/dist-packages/pytools/__init__.py:519: KeyError

During handling of the above exception, another exception occurred:

self = <compyle.tests.test_parallel.TestParallelUtilsJIT testMethod=test_repeated_scans_with_different_settings>

    def test_repeated_scans_with_different_settings(self):
        importorskip('pyopencl')
        with use_config(use_double=False):
>           self._test_unique_scan(backend='opencl')

compyle/tests/test_parallel.py:197: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compyle/tests/test_parallel.py:753: in _test_unique_scan
    scan(ary=a, unique=unique_ary, unique_count=unique_count)
compyle/parallel.py:1242: in __call__
    self.scan(**kwargs)
compyle/profile.py:48: in wrapper
    return method(*args, **kwargs)
compyle/jit.py:522: in __call__
    c_func = self._generate_kernel(**kwargs)
<decorator-gen-12>:2: in _generate_kernel
    ???
/usr/lib/python3/dist-packages/pytools/__init__.py:526: in _deco
    result = func(*args, **kwargs)
compyle/jit.py:510: in _generate_kernel
    return self._generate(declarations=declarations)
compyle/parallel.py:902: in _generate
    return self._generate_opencl_kernel(declarations=declarations)
compyle/parallel.py:1087: in _generate_opencl_kernel
    segment_expr, preamble = self._get_opencl_cuda_code(
compyle/parallel.py:1057: in _get_opencl_cuda_code
    self._wrap_ocl_function(self.input_func, func_type='input',
compyle/parallel.py:1022: in _wrap_ocl_function
    self.tp.add(func, declarations=declarations)
compyle/transpiler.py:278: in add
    code = self._cgen.parse(
compyle/translator.py:281: in parse
    code = self.parse_function(obj, declarations=declarations)
compyle/translator.py:307: in parse_function
    code = self.convert(src)
compyle/translator.py:252: in convert
    result = self.visit(code)
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:652: in visit_Module
    return '\n'.join(
compyle/translator.py:653: in <genexpr>
    self.visit(item) for item in node.body
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:584: in visit_FunctionDef
    body = '\n'.join(self._indent_block(self.visit(item))
compyle/translator.py:584: in <genexpr>
    body = '\n'.join(self._indent_block(self.visit(item))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:615: in visit_If
    cond=self.visit(node.test),
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:376: in visit_BoolOp
    return '(%s)' % (op.join(self.visit(x) for x in node.values))
compyle/translator.py:376: in <genexpr>
    return '(%s)' % (op.join(self.visit(x) for x in node.values))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:426: in visit_Compare
    self.visit(node.comparators[0]))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <compyle.translator.OpenCLConverter object at 0x7fd66bcf2160>
node = <ast.Subscript object at 0x7fd66baf8760>

    def visit_Subscript(self, node):
        if isinstance(node.slice, ast.Constant) or isinstance(node.slice, ast.Name):
            return '%s[%s]' % (
                self.visit(node.value), self.visit(node.slice)
            )
        else:
            return '%s[%s]' % (
>               self.visit(node.value), self.visit(node.slice.value)
            )
E           AttributeError: 'BinOp' object has no attribute 'value'

compyle/translator.py:709: AttributeError
_________________ TestParallelUtilsJIT.test_unique_scan_opencl _________________

func = <function ScanJIT._generate_kernel at 0x7fd68bf06430>
args = (<compyle.jit.ScanJIT object at 0x7fd66bc52d90>,)
kwargs = {'ary': <compyle.array.Array object at 0x7fd66bc52f70>, 'unique': <compyle.array.Array object at 0x7fd66bc527f0>, 'unique_count': <compyle.array.Array object at 0x7fd66bc523a0>}
key = ('gintp', 'gintp', 'gintp', <function TestParallelUtilsJIT._test_unique_scan.<locals>.input_f at 0x7fd66bc8ed30>, <function TestParallelUtilsJIT._test_unique_scan.<locals>.output_f at 0x7fd66b3f4dc0>, 'a+b', ...)

    @my_decorator
    def _deco(func, *args, **kwargs):
        # by Michele Simionato
        # http://www.phyast.pitt.edu/~micheles/python/
        key = key_func(*args, **kwargs)
        try:
>           return func._memoize_dic[key]  # pylint: disable=protected-access
E           KeyError: ('gintp', 'gintp', 'gintp', <function TestParallelUtilsJIT._test_unique_scan.<locals>.input_f at 0x7fd66bc8ed30>, <function TestParallelUtilsJIT._test_unique_scan.<locals>.output_f at 0x7fd66b3f4dc0>, 'a+b', 'opencl', False, True)

/usr/lib/python3/dist-packages/pytools/__init__.py:519: KeyError

During handling of the above exception, another exception occurred:

self = <compyle.tests.test_parallel.TestParallelUtilsJIT testMethod=test_unique_scan_opencl>

    def test_unique_scan_opencl(self):
        importorskip('pyopencl')
>       self._test_unique_scan(backend='opencl')

compyle/tests/test_parallel.py:134: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
compyle/tests/test_parallel.py:753: in _test_unique_scan
    scan(ary=a, unique=unique_ary, unique_count=unique_count)
compyle/parallel.py:1242: in __call__
    self.scan(**kwargs)
compyle/profile.py:48: in wrapper
    return method(*args, **kwargs)
compyle/jit.py:522: in __call__
    c_func = self._generate_kernel(**kwargs)
<decorator-gen-12>:2: in _generate_kernel
    ???
/usr/lib/python3/dist-packages/pytools/__init__.py:526: in _deco
    result = func(*args, **kwargs)
compyle/jit.py:510: in _generate_kernel
    return self._generate(declarations=declarations)
compyle/parallel.py:902: in _generate
    return self._generate_opencl_kernel(declarations=declarations)
compyle/parallel.py:1087: in _generate_opencl_kernel
    segment_expr, preamble = self._get_opencl_cuda_code(
compyle/parallel.py:1057: in _get_opencl_cuda_code
    self._wrap_ocl_function(self.input_func, func_type='input',
compyle/parallel.py:1022: in _wrap_ocl_function
    self.tp.add(func, declarations=declarations)
compyle/transpiler.py:278: in add
    code = self._cgen.parse(
compyle/translator.py:281: in parse
    code = self.parse_function(obj, declarations=declarations)
compyle/translator.py:307: in parse_function
    code = self.convert(src)
compyle/translator.py:252: in convert
    result = self.visit(code)
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:652: in visit_Module
    return '\n'.join(
compyle/translator.py:653: in <genexpr>
    self.visit(item) for item in node.body
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:584: in visit_FunctionDef
    body = '\n'.join(self._indent_block(self.visit(item))
compyle/translator.py:584: in <genexpr>
    body = '\n'.join(self._indent_block(self.visit(item))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:615: in visit_If
    cond=self.visit(node.test),
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:376: in visit_BoolOp
    return '(%s)' % (op.join(self.visit(x) for x in node.values))
compyle/translator.py:376: in <genexpr>
    return '(%s)' % (op.join(self.visit(x) for x in node.values))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
compyle/translator.py:426: in visit_Compare
    self.visit(node.comparators[0]))
/usr/lib/python3.9/ast.py:407: in visit
    return visitor(node)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <compyle.translator.OpenCLConverter object at 0x7fd66bc52760>
node = <ast.Subscript object at 0x7fd66b984640>

    def visit_Subscript(self, node):
        if isinstance(node.slice, ast.Constant) or isinstance(node.slice, ast.Name):
            return '%s[%s]' % (
                self.visit(node.value), self.visit(node.slice)
            )
        else:
            return '%s[%s]' % (
>               self.visit(node.value), self.visit(node.slice.value)
            )
E           AttributeError: 'BinOp' object has no attribute 'value'

compyle/translator.py:709: AttributeError
=============================== warnings summary ===============================

[CUT]

-- Docs: https://docs.pytest.org/en/latest/warnings.html
= 7 failed, 241 passed, 43 skipped, 4 deselected, 156 warnings in 886.44 seconds =
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -8,6 +8,7 @@ language: python
 python:
     - 3.7
     - 3.8
+    - 3.9
 
 install:
     - sudo apt-get update

--- a/compyle/translator.py
+++ b/compyle/translator.py
@@ -700,8 +700,12 @@ class CConverter(ast.NodeVisitor):
         return r'"%s"' % node.s
 
     def visit_Subscript(self, node):
+        if sys.version_info >= (3, 9):
+            sliceval = node.slice
+        else:
+            sliceval = node.slice.value
         return '%s[%s]' % (
-            self.visit(node.value), self.visit(node.slice.value)
+            self.visit(node.value), self.visit(sliceval)
         )
 
     def visit_TryExcept(self, node):

